include/irq.h

```diff
@@ -50,6 +50,10 @@ extern "C" {
 #define IRQ_CONNECT(irq_p, priority_p, isr_p, isr_param_p, flags_p) \
 	Z_ARCH_IRQ_CONNECT(irq_p, priority_p, isr_p, isr_param_p, flags_p)
 
+extern int z_arch_irq_connect_dynamic(unsigned int irq, unsigned int priority,
```

Comment:
Maybe a stupid comment, but, is this declaration meant only for `irq_connect_dynamic`? or is it a public interface?
If it is only for `irq_connect_dynamic` maybe you want to hide this inside `irq_connect_dynamic`'s body(?)

Reply:
Oh, I moved this so that the doxygen was associated with the right function, otherwise we get a doc error.

Eventually all of the z_arch_* prototypes found scattered across the tree will be removed in favor of the definitions in these new headers, there's other PRs that have to land before I can do that, see #19594 

A combined PR with *all* the changes is a huge code change (see https://github.com/andrewboie/zephyr/tree/arch-interface-docs-wiring) , I'm attempting to split this into more easy-to-review chunks.